trieste/models/interfaces.py

```diff
@@ -241,6 +264,113 @@ def log(self) -> None:
             with tf.name_scope(f"{i}"):
                 model.log()
 
+    def reparam_sampler(self, num_samples: int) -> ReparametrizationSampler:
+        """
+        Return a reparameterization sampler providing `num_samples` samples across
+        all the models in the model stack. This is currently only implemented for
+        stacks made from models that have a :class:`BatchReparametrizationSampler`
+        as their reparameterization sampler.
+
+        :param num_samples: The desired number of samples.
+        :return: The reparametrization sampler.
+        :raise NotImplementedError: If the models in the stack do not share the
+            same :meth:`reparam_sampler`.
+        """
+
+        sampler_types = [type(model.reparam_sampler(1)) for model in self._models]
+        unique_sampler_types = set(sampler_types)
```

Comment:
I would still simplify this and prepare it for when/if we support heterogeneous sampler types:

```
samplers = [model.reparam_sampler(num_samples) for model in self._models]
unique_sampler_types = set(type(sampler) for sampler in sampler_types)
if len(unique_sampler_types) == 1:
    # currently assume that all sampler constructors look the same
    shared_sampler_type = type(samplers[0])
    return shared_sampler_type(num_samples, self)
else:
    raise NotImplementedError(...)
```